File: src/SIL.Machine.TS/src/translation/ecm-score-info.ts
import { EditOperation } from './edit-operation';

export class EcmScoreInfo {
  readonly scores: number[] = [];
  readonly operations: EditOperation[] = [];

  updatePositions(prevEsi: EcmScoreInfo, positions: number[]): void {
    while (this.scores.length < prevEsi.scores.length) {
      this.scores.push(0);
    }

    while (this.operations.length < prevEsi.operations.length) {
      this.operations.push(EditOperation.None);
    }

    for (let i = 0; i < positions.length; i++) {
      this.scores[positions[i]] = prevEsi.scores[positions[i]];
      if (prevEsi.operations.length > i) {
        this.operations[positions[i]] = prevEsi.operations[positions[i]];
      }
    }
  }

  removeLast(): void {
    if (this.scores.length > 1) {
      this.scores.pop();
    }
    if (this.operations.length > 1) {
      this.operations.pop();
    }
  }

  getLastInsPrefixWordFromEsi(): number[] {
    const results = new Array<number>(this.operations.length);

    for (let j = this.operations.length - 1; j >= 0; j--) {
      switch (this.operations[j]) {
        case EditOperation.Hit:
          results[j] = j - 1;
          break;

        case EditOperation.Insert:
          let tj = j;
          while (tj >= 0 && this.operations[tj] === EditOperation.Insert) {
            tj--;
          }
          if (this.operations[tj] === EditOperation.Hit || this.operations[tj] === EditOperation.Substitute) {
            tj--;
          }
          results[j] = tj;
          break;

        case EditOperation.Delete:
          results[j] = j;
          break;

        case EditOperation.Substitute:
          results[j] = j - 1;
          break;

        case EditOperation.None:
          results[j] = 0;
          break;
      }
    }

    return results;
  }
}

File: src/SIL.Machine/Translation/EcmScoreInfo.cs
using System.Collections.Generic;

namespace SIL.Machine.Translation
{
    public class EcmScoreInfo
    {
        public List<double> Scores { get; } = new List<double>();
        public List<EditOperation> Operations { get; } = new List<EditOperation>();

        public void UpdatePositions(EcmScoreInfo prevEsi, List<int> positions)
        {
            while (Scores.Count < prevEsi.Scores.Count)
                Scores.Add(0);

            while (Operations.Count < prevEsi.Operations.Count)
                Operations.Add(EditOperation.None);

            for (int i = 0; i < positions.Count; i++)
            {
                Scores[positions[i]] = prevEsi.Scores[positions[i]];
                if (prevEsi.Operations.Count > i)
                    Operations[positions[i]] = prevEsi.Operations[positions[i]];
            }
        }

        public void RemoveLast()
        {
            if (Scores.Count > 1)
                Scores.RemoveAt(Scores.Count - 1);
            if (Operations.Count > 1)
                Operations.RemoveAt(Operations.Count - 1);
        }

        public int[] GetLastInsPrefixWordFromEsi()
        {
            var results = new int[Operations.Count];

            for (int j = Operations.Count - 1; j >= 0; j--)
            {
                switch (Operations[j])
                {
                    case EditOperation.Hit:
                        results[j] = j - 1;
                        break;

                    case EditOperation.Insert:
                        int tj = j;
                        while (tj >= 0 && Operations[tj] == EditOperation.Insert)
                            tj--;
                        if (Operations[tj] == EditOperation.Hit || Operations[tj] == EditOperation.Substitute)
                            tj--;
                        results[j] = tj;
                        break;

                    case EditOperation.Delete:
                        results[j] = j;
                        break;

                    case EditOperation.Substitute:
                        results[j] = j - 1;
                        break;

                    case EditOperation.None:
                        results[j] = 0;
                        break;
                }
            }

            return results;
        }
    }
}
